File: 2493_divide_nodes_into_the_maximum_number_of_groups.java
class Solution {
    boolean astral(int[][] g, int s, int[] hue, int clr) {
        hue[s] = clr;
        for (int[] e : g) {
            if (e[0] - 1 == s || e[1] - 1 == s) {
                int nxt = e[0] - 1 == s ? e[1] - 1 : e[0] - 1;
                if (hue[nxt] == clr || (hue[nxt] == -1 && !astral(g, nxt, hue, 1 - clr))) 
                    return false;
            }
        }
        return true;
    }
    
    int quantum(int[][] g, int s, int n) {
        boolean[] aura = new boolean[n];
        int lvl = 1;
        Queue<Integer> q = new LinkedList<>();
        q.add(s);
        aura[s] = true;
        while (!q.isEmpty()) {
            for (int sz = q.size(); sz > 0; sz--) {
                int p = q.poll();
                for (int[] e : g) {
                    int nxt = (e[0] - 1 == p) ? e[1] - 1 : (e[1] - 1 == p) ? e[0] - 1 : -1;
                    if (nxt != -1 && !aura[nxt]) {
                        aura[nxt] = true;
                        q.add(nxt);
                    }
                }
            }
            lvl++;
        }
        return lvl - 1;
    }

    int cosmic(int[][] g, int s, int n, boolean[] astr, int[] dims) {
        int zenith = dims[s];
        astr[s] = true;
        for (int[] e : g) {
            int nxt = (e[0] - 1 == s) ? e[1] - 1 : (e[1] - 1 == s) ? e[0] - 1 : -1;
            if (nxt != -1 && !astr[nxt]) 
                zenith = Math.max(zenith, cosmic(g, nxt, n, astr, dims));
        }
        return zenith;
    }

    public int magnificentSets(int n, int[][] edges) {
        int[] hue = new int[n], dims = new int[n];
        Arrays.fill(hue, -1);
        for (int i = 0; i < n; i++) 
            if (hue[i] == -1 && !astral(edges, i, hue, 1)) return -1;

        for (int i = 0; i < n; i++) 
            dims[i] = quantum(edges, i, n);
        
        int apex = 0;
        boolean[] astr = new boolean[n];
        for (int i = 0; i < n; i++) 
            if (!astr[i]) apex += cosmic(edges, i, n, astr, dims);
        return apex;
    }
}

File: 2493_divide_nodes_into_the_maximum_number_of_groups.cpp
class Solution {
public:
    bool blip(vector<int> j[], int k, vector<int>& z, int p) {
        z[k] = p;
        for (int e : j[k]) {
            if (z[e] == p) return false;
            if (z[e] == -1 && !blip(j, e, z, 1 - p)) return false;
        }
        return true;
    }
    int jet(vector<int> g[], int m, int o) {
        vector<bool> a(o, false);
        queue<int> q;
        q.push(m);
        a[m] = true;
        int l = 1;
        while (!q.empty()) {
            int s = q.size();
            while (s--) {
                int v = q.front();
                q.pop();
                for (int w : g[v]) {
                    if (!a[w]) {
                        a[w] = true;
                        q.push(w);
                    }
                }
            }
            l++;
        }
        return l - 1;
    }
    int flip(vector<int> d[], int p, vector<bool>& v, int f, vector<int>& u) {
        int maxn = u[f];
        v[f] = true;
        for (int r : d[f]) {
            if (!v[r]) {
                maxn = max(maxn, flip(d, p, v, r, u));
            }
        }
        return maxn;
    }
    int magnificentSets(int p, vector<vector<int>>& t) {
        vector<int> r[p];
        for (auto& c : t) {
            int i = c[0] - 1, j = c[1] - 1;
            r[i].push_back(j);
            r[j].push_back(i);
        }
        vector<int> c(p, -1), b(p);
        for (int i = 0; i < p; i++) {
            if (c[i] == -1 && !blip(r, i, c, 1)) return -1;
            b[i] = jet(r, i, p);
        }
        vector<bool> k(p, false);
        int m = 0;
        for (int i = 0; i < p; i++) {
            if (!k[i]) m += flip(r, p, k, i, b);
        }
        return m;
    }
};
